library/ssl_srv.c

```diff
@@ -2765,6 +2765,53 @@ static int ssl_write_hello_verify_request( mbedtls_ssl_context *ssl )
 }
 #endif /* MBEDTLS_SSL_DTLS_HELLO_VERIFY */
 
+static void ssl_check_id_based_session_resumption( mbedtls_ssl_context *ssl )
+{
+    int ret;
+    mbedtls_ssl_session session_tmp = { 0 };
+    mbedtls_ssl_session * const session = ssl->session_negotiate;
+
+    /* Resume is 0  by default, see ssl_handshake_init().
+     * It may be already set to 1 by ssl_parse_session_ticket_ext(). */
+    if( ssl->handshake->resume == 1 )
+        return;
+    if( ssl->session_negotiate->id_len == 0 )
+        return;
+    if( ssl->conf->f_get_cache == NULL )
+        return;
+#if defined(MBEDTLS_SSL_RENEGOTIATION)
+    if( ssl->renego_status != MBEDTLS_SSL_INITIAL_HANDSHAKE )
+        return;
+#endif
+
+    session_tmp.id_len = session->id_len;
+    memcpy( session_tmp.id, session->id, session->id_len );
+
+    ret = ssl->conf->f_get_cache( ssl->conf->p_cache,
+                                  &session_tmp );
+    if( ret != 0 )
+        goto exit;
+
+    if( session->ciphersuite != session_tmp.ciphersuite ||
+        session->compression != session_tmp.compression )
+    {
+        /* Mismatch between cached and negotiated session */
+        goto exit;
+    }
+
+    /* Move semantics */
+    /* Zeroization of session_tmp happens at the end of the function. */
+    mbedtls_ssl_session_free( session );
+    *session = session_tmp;
+
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "session successfully restored from cache" ) );
+    ssl->handshake->resume = 1;
+
+exit:
+
+    mbedtls_platform_zeroize( &session_tmp, sizeof( session_tmp ) );
```

Comment:
This works well if the callback works atomically (always returns an empty session on error), but if the callback is of less than optimal quality, we might get an error code with a loaded or half-baked session that needs deep freeing to avoid resource leaks.

Reply:
Note that this has been changed to `mbedtls_ssl_session_free()`. Regarding half-baked structures, I think there's nothing we can do here. We have to assume -- but perhaps better document! -- that the session cache callback must always return a session instance which can be safely passed to `mbedtls_ssl_session_free()`, whether it fails or succeeds.

Reply:
And indeed it wasn't properly documented. I have added a commit to improve the documentation of `mbedtls_ssl_cache_get_t()`.

Reply:
https://github.com/ARMmbed/mbedtls/pull/4347/commits/0d05f40222a513a9dcce937583f9de337aa0659f